int32 p
int32 i
int32 d
int16 i_max
